Chart: custom context menu

cnrudd opened the pull request:
This addresses the following tickets:

https://github.com/xh/hoist-react/issues/2389
https://github.com/xh/hoist-react/issues/3813
https://github.com/xh/hoist-react/issues/3388
https://github.com/xh/hoist-react/issues/2753

Yes, you get 4 for 1 by merging this PR.  :)

Corresponding Toolbox PR with new custom context menu on Line Chart: https://github.com/xh/toolbox/pull/765

The scope of this PR has expanded.  The PR now proposes these changes:

1. New ability to customize chart context menus with a preset list of chart menu tokens or custom menu items.
2. Custom chart menu items will receive the active chart series points and the original right click event as args.
3. The framework's standard context menu's menu item actionFns will also receive the original right click event as a 2nd arg



Hoist P/R Checklist
-------------------

**Pull request authors:** Review and check off the below. Items that do not apply can also be
checked off to indicate they have been considered. If unclear if a step is relevant, please leave
unchecked and note in comments.

- [x] Caught up with `develop` branch as of last change.
- [x] Added CHANGELOG entry, or determined not required.
- [x] Reviewed for breaking changes: minor breaking change, noted in changelog.
- [x] Updated doc comments / prop-types, or determined not required.
- [x] Reviewed and tested on Mobile: not required.  no context menu on mobile
- [x] Created Toolbox branch: https://github.com/xh/toolbox/pull/765

**If your change is still a WIP**, please use the "Create draft pull request" option in the split
button below to indicate it is not ready yet for a final review.

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

amcclain commented:
Thanks for this @cnrudd  - I will review this very shortly and reach out with any questions

amcclain commented on desktop/cmp/contextmenu/ChartContextMenu.ts:
```diff
@@ -0,0 +1,194 @@
+        this.contextMenuClickEvt = contextMenuClickEvt;
+        this.point = point;
+        this.items = cloneDeep(items).map(it => this.buildMenuItemConfig(it));
+    }
```
It took me a little while reading this PR to grok, but realizing that this is a funny class / object to have.  We create it and it captures this state in its ctor, but then all we do is read its `items` property and then never reference it again (with this all happening in the `ChartLocalModel`.

Give that, it seems to me that this would be better implemented as a pure function - that's really what we need to deliver - a function that takes an event from HC and returns the items in the form that HC is ready to consume.  

Having the object right now made me immediately think it was more stateful than it really is. It also gave us another thing to keep track of in our minds when we say "contextMenu" - we have this object, the spec on `ChartModel`, and the function to always produce a new one of these objects on `ChartLocalModel`.

amcclain reviewed (commented):
This was pre-existing code, but since we're looking at the context menu here, in `Chart.ts` render function we have

`return !XH.isMobileApp ? useContextMenu(coreContents, impl.contextMenu) : coreContents;`

Seems like we should check `impl.contextMenu` instead and return `coreContents` directly if null.  That way the "not supported on mobile" fact is expressed once, and we also aren't wrapping the component in the case where the user has disabled the context menu.  

------

Thanks for making this change - I think we have all the parts we need, just wondering in my review comments here about the "packaging" - specifically the roles of `ChartModel` vs. `ChartLocalModel` and the need for the new `ChartContextMenu` class vs a functional implementation.  Have a look and LMK when you think when you have a chance.

cnrudd commented:
@amcclain @lbwexler This is ready for a 2nd look.

I've implemented these changes:

1. reworked as @amcclain suggestion so that menu item building is no longer in a class (and a lot of the smaller suggestions)
2. added contextMenuEvent to standard context menu (Lee's suggestion)
3. chart actionFn: grouped args in 'params' arg (Lee's suggestion)
4. chart actionFn: support providing points (for when more than one series is in chart and tooltip is shared)
5. also fixed chart mouse right to left zoom out bug
6. added/improved toolbox demos: multi series chart, new context menu item on basicPanel demo

amcclain commented on cmp/chart/impl/ChartContextMenuItems.ts:
```diff
@@ -0,0 +1,201 @@
+
+/**
+ * If a String, value can be '-' for a separator, or a token supported by HighCharts
+ * for its native menu items, or a Hoist specific token.
```
Can it also be any arbitrary string to insert a header?  If not, think we should spec `'-'` vs `string` here

amcclain commented on cmp/chart/impl/ChartContextMenuItems.ts:
```diff
@@ -0,0 +1,201 @@
+            return {
+                text: 'View in full screen',
+                icon: Icon.expand(),
+                hidden: !chartModel,
```
I notice that all of these have this `hidden: !chartModel` - are we ever rendering this menu without a `chartModel`?  Should there be some early out before we reach this point?

amcclain commented on cmp/chart/impl/ChartContextMenuItems.ts:
```diff
@@ -0,0 +1,201 @@
+    | 'downloadPDF'
+    | 'copyToClipboard';
+
+export interface ChartMenuItem extends Omit<MenuItem, 'actionFn' | 'items'> {
```
Worth making the naming consistent - we could drop `Context` from the other exported members here (there's no other kind of menu), or make this `ChartContextMenuItem`

amcclain commented on cmp/chart/impl/ChartContextMenuItems.ts:
```diff
@@ -0,0 +1,201 @@
+}
+
+function isMenuItem(item: ChartMenuItemLike): item is MenuItem {
+    return !isString(item) && !isValidElement(item);
```
This is called in just one place here, and we've already checked if the value is a string and won't reach here if so - could inline the validElement check and drop this function which I think just adds a bit more noise to the file at the moment

amcclain commented on cmp/chart/impl/zoomout.ts:
```diff
@@ -30,7 +30,8 @@ export function installZoomoutGesture(Highcharts) {
         Highcharts.addEvent(this, 'selection', e => {
             if (pixelDiff < 0) {
-                this.zoom(); // call w/o arguments resets zoom
+                // `undefined`'s preserve defaults, last arg `false` disables animation
```
What's this change doing? Doesn't seem related to the others, or is it?

amcclain commented on core/types/Interfaces.ts:
```diff
@@ -284,10 +284,10 @@ export interface MenuItem {
     /** Executed when the user clicks the menu item. */
-    actionFn?: (e: MouseEvent | PointerEvent) => void;
+    actionFn?: (e: MouseEvent | PointerEvent, contextMenuEvent?: MouseEvent | PointerEvent) => void;
 
```
So the first event `e` is the click on a specific item within the menu, and the new second event `contextMenuEvent` is the right-click that caused the menu to appear in the first place?

amcclain commented on desktop/cmp/contextmenu/ContextMenu.ts:
```diff
@@ -20,6 +20,12 @@ export type ContextMenuSpec = MenuItemLike[] | ((e: MouseEvent) => MenuItemLike[
     menuItems: MenuItemLike[];
+    /**
+     * Optional event because this component is sometimes used as the content in popovers, to
+     * create menu buttons.  In those cases, it is not expected that the click event contains any
```
> because this component is sometimes used as the content in popovers, to create menu buttons

Is this not *always* how it's used?  What's an example of the "contextual information" here, or when would you expect to find something in the first event `e`?  Just as a developer I'm left a little uncertain as to what I would do with one event or the other.

amcclain commented on cmp/chart/impl/ChartContextMenuItems.ts:
```diff
@@ -0,0 +1,201 @@
+    | ChartContextMenuItemLike[]
+    | ((chartModel: ChartModel) => ChartContextMenuItemLike[]);
+
+export function getChartContextMenuItems(
```
Should we mark as `@internal`?

amcclain commented on cmp/chart/ChartModel.ts:
```diff
@@ -95,4 +111,13 @@ export class ChartModel extends HoistModel {
+        if (isNil(spec) || spec === true) return ChartModel.defaultContextMenu;
+
+        return spec;
+    }
```
This looks great / very clear

amcclain commented on cmp/chart/Chart.ts:
```diff
@@ -106,7 +108,7 @@ class ChartLocalModel extends HoistModel {
 
     chartRef = createObservableRef<HTMLElement>();
-    contextMenu: any;
+    contextMenu: (e) => MenuItemLike[];
```
Sorry we can leave this as is, but still feels like there's some way to streamline a bit more.

Currently we have
* `ChartLocalModel.contextMenu` - function created/captured as soon as linked
* `ChartLocalModel.getContextMenu()` - function to create the above function
* `ChartModel.contextMenu` - partially processed/parsed spec, might be a function that this local model calls, passing back reference to `ChartModel` itself
* `getChartContextMenuItems()` exported util

Still wondering what we're getting from having the context menu stuff in this local model - I am probably missing some subtlety as to why we need this layer involved, but whatever that subtlety is, if we could make more clear might help whenever we revisit this code in the future.

Or maybe we could drop contextMenu from this local model altogether and have the primary `ChartModel` process the specs/tokens it's given "all the way" into the function that `ChartLocalModel.getContextMenu()` currently produces.  

So we'd have context menu stuff in two places - the main `ChartModel` and the helper util - and `ChartModel` in its ctor would process the specs it's passed all the way through to the form that the rendering within `Chart.ts` consumes and passes on to the `useContextMenu` hook.

Don't mean to be a pain with all of this, but am having a hard time keeping track of the current flow and if it could be simpler, great - if not, no worries :)

lbwexler commented on cmp/chart/Chart.ts:
```diff
@@ -95,7 +94,8 @@ export const [Chart, chart] = hoistCmp.withFactory<ChartProps>({
 
-        return !XH.isMobileApp ? useContextMenu(coreContents, impl.contextMenu) : coreContents;
+        // Must check isMobileApp here because `useContextMenu` is not defined in mobile
+        return !XH.isMobileApp ? useContextMenu(coreContents, model.contextMenu) : coreContents;
```
👍🏻

lbwexler reviewed (commented):
Thanks for taking this one step further Colin.  Just a couple of small tweaks and hope we can merge this AM

lbwexler commented on CHANGELOG.md:
```diff
@@ -10,12 +10,16 @@
 * Added `ViewManagerModel.preserveUnsavedChanges` flag to opt-out of that behaviour.
 * Added `PersistOptions.settleTime` to configure time to wait for state to settle before persisting.
 * Support for gridcolumn level `onCellClicked` events.
+* Chart context menu is now fully customizable.
```
We could mention in the context of "general improvements to menu api"  and also mention new MenuContext object now sent generally to both handlers as second argument.
